cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetSpecificationUpgrader.java

```diff
@@ -0,0 +1,137 @@
+   * </p>
+   * @throws Exception
+   */
+  public void upgrade() throws Exception {
+    TableId datasetSpecId = TableId.from(Id.Namespace.SYSTEM.getId(), DatasetMetaTableUtil.INSTANCE_TABLE_NAME);
+    HBaseAdmin hBaseAdmin = new HBaseAdmin(conf);
+    if (!tableUtil.tableExists(hBaseAdmin, datasetSpecId)) {
+      LOG.info("Dataset instance table does not exist: {}. Should not happen", datasetSpecId);
+      return;
+    }
+
+    HTable specTable = tableUtil.createHTable(conf, datasetSpecId);
+
+    try {
+      ScanBuilder scanBuilder = tableUtil.buildScan();
+      scanBuilder.setTimeRange(0, HConstants.LATEST_TIMESTAMP);
+      scanBuilder.setMaxVersions();
+      try (ResultScanner resultScanner = specTable.getScanner(scanBuilder.build())) {
+        Result result;
+        while ((result = resultScanner.next()) != null) {
+          Put put = new Put(result.getRow());
+          for (Map.Entry<byte[], NavigableMap<byte[], NavigableMap<Long, byte[]>>> familyMap :
+            result.getMap().entrySet()) {
+            for (Map.Entry<byte[], NavigableMap<Long, byte[]>> columnMap : familyMap.getValue().entrySet()) {
+              for (Map.Entry<Long, byte[]> columnEntry : columnMap.getValue().entrySet()) {
+                Long timeStamp = columnEntry.getKey();
+                byte[] colVal = columnEntry.getValue();
+                String specEntry = Bytes.toString(colVal);
+                DatasetSpecification specification = GSON.fromJson(specEntry, DatasetSpecification.class);
+                DatasetSpecification updatedSpec = updateTTLInSpecification(specification);
+                colVal = Bytes.toBytes(GSON.toJson(updatedSpec));
+                put.add(familyMap.getKey(), columnMap.getKey(), timeStamp, colVal);
+              }
+            }
+          }
+          specTable.put(put);
+        }
+      }
+    } finally {
+      specTable.flushCommits();
+      specTable.close();
+    }
+  }
+
+  private Map<String, String> updatedProperties(Map<String, String> properties) {
+    if (properties.containsKey(Table.PROPERTY_TTL)) {
+      SortedMap<String, String> updatedProperties = new TreeMap<>(properties);
+      long updatedValue = TimeUnit.MILLISECONDS.toSeconds(Long.valueOf(updatedProperties.get(Table.PROPERTY_TTL)));
+      updatedProperties.put(Table.PROPERTY_TTL, String.valueOf(updatedValue));
+      return updatedProperties;
+    }
+    return properties;
+  }
+
+  @VisibleForTesting
+  DatasetSpecification updateTTLInSpecification(DatasetSpecification specification) {
+    Map<String, String> properties = updatedProperties(specification.getProperties());
+    List<DatasetSpecification> updatedSpecs = new ArrayList<>();
+    for (DatasetSpecification datasetSpecification : specification.getSpecifications().values()) {
+      updatedSpecs.add(updateTTLInSpecification(datasetSpecification));
+    }
+    return DatasetSpecification.builder(specification.getName(),
```

Comment:
I believe this is wrong. Because spec.getName will be enclosingName.name. You need to use the key from the specs map that has the unqualified name. Better double-check.